### DISCUSSION TITLE
Problem with SoftwareSerial attiny861

### DISCUSSION THREAD

**Discussion by DavidAlonsoGalindo** [SpenceKonde/ATTinyCore#571] [Support / Q & A]
Hi spence konde

Currently I am working with the attiny861 and I would like to have two serial communications one to communicate through an FTDI to the computer and the other to communicate with an attiny85, but I have a problem when using the serial Software library, when loading the project the attiny861 it restarts at every moment, with the Serial that comes by default it works very well, but by including the serial software it starts the problem.

Could you advise me how to solve the problem?

**Comment by SpenceKonde** [comment]
I'll bet I know what's causing it.....  If you enable warnings, do you get an warning about a missspelled vector?

**Reply by SpenceKonde** [reply]
the 861 is unique among like... everything.... in having 2 banks of =PCINTs, but only one PCINT vector instead of PCINT0 and PCINT1. I'd wager that the software serial library is getting this wrong. Bizzarely, misspelling an ISR name is a warning not an error.

**Comment by DavidAlonsoGalindo** [comment]
the configuration of the pinout is the new style, the example code that I load in the arduino is the following:

int c;

#include <SoftwareSerial.h>

const int rx=0;
const int tx=1;

SoftwareSerial  portOne(rx,tx);
void setup()
{
   pinMode(rx,INPUT);
  pinMode(tx,OUTPUT);
  Serial.begin(9600);
  portOne.begin(9600);
  
}

void loop()
{

 Serial.print("C: ");
 Serial.println(c);
 delay(1000);
 c++;
}

**Reply by DavidAlonsoGalindo** [reply]
When compiling the program the warnings are:

C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp: In function 'void DebugPulse(uint8_t, uint8_t)':
C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp:60:32: warning: unused parameter 'pin' [-Wunused-parameter]
 inline void DebugPulse(uint8_t pin, uint8_t count)
                                ^~~
C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp:60:45: warning: unused parameter 'count' [-Wunused-parameter]
 inline void DebugPulse(uint8_t pin, uint8_t count)
                                             ^~~~~
C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp: At global scope:
C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp:375:6: warning: always_inline function might not be inlinable [-Wattributes]
 void SoftwareSerial::setRxIntMsk(bool enable)
      ^~~~~~~~~~~~~~
C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp:121:6: warning: always_inline function might not be inlinable [-Wattributes]
 void SoftwareSerial::recv()
      ^~~~~~~~~~~~~~




when I load the bootloader the warnings are:

 ***failed;  
avrdude: WARNING: invalid value for unused bits in fuse "efuse", should be set to 1 according to datasheet
This behaviour is deprecated and will result in an error in future version
You probably want to use 0x01 instead of 0xff (double check with your datasheet first).

**Comment by DavidAlonsoGalindo** [comment]
once the program is loaded into the attiny861
and verify the serial monitor the following data appears:

CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0
CC: 0C:C: 0

**Comment by SpenceKonde** [comment]
no ISR warning? So they handled PCINT vectors correctly here? *checks*... oh..... well that would do it...

From the library:
```
#if defined(PCINT0_vect)
ISR(PCINT0_vect)
{
  SoftwareSerial::handle_interrupt();
}
#endif

#if defined(PCINT1_vect)
ISR(PCINT1_vect, ISR_ALIASOF(PCINT0_vect));
#endif

#if defined(PCINT2_vect)
ISR(PCINT2_vect, ISR_ALIASOF(PCINT0_vect));
#endif

#if defined(PCINT3_vect)
ISR(PCINT3_vect, ISR_ALIASOF(PCINT0_vect));
#endif
```

From the io headers:
```
#define PCINT_vect		    	_VECTOR(2)
```

Library doesn't look for PCINT_vect, only PCINT0/1/2/3. *and it doesn't complain when it doesn't find any ISR* (god softwareserial is a terrible library, I hate everything about it!)
So it does not define any the interrupt handler, but successfully configures the interrupt to fire. Interrupt fires when the serial pin changes. There is no interrupt defined so avr-gcc defaults to it's BAD_ISR routine. Which jumps to the reset vector. Repeat.

**Comment by SpenceKonde** [comment]
Adding 
```
#if defined(PCINT_vect)
ISR(PCINT_vect)
{
  SoftwareSerial::handle_interrupt();
}
#endif
```
to that part of SoftwareSerial.cpp will fix it. Can you please make an issue about this so I remember to get the fix into the next release?

**Reply by DavidAlonsoGalindo** [reply]
Yes of course

**Comment by DavidAlonsoGalindo** [comment]
an apology I recompiled the program and now I got these warnings, this happened when changing the pinout from new style to legacy
When compiling the program the warnings are:

C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp: In function 'void DebugPulse(uint8_t, uint8_t)':
C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp:60:32: warning: unused parameter 'pin' [-Wunused-parameter]
inline void DebugPulse(uint8_t pin, uint8_t count)
^~~
C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp:60:45: warning: unused parameter 'count' [-Wunused-parameter]
inline void DebugPulse(uint8_t pin, uint8_t count)
^~~~~
C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp: At global scope:
C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp:375:6: warning: always_inline function might not be inlinable [-Wattributes]
void SoftwareSerial::setRxIntMsk(bool enable)
^~~~~~~~~~~~~~
C:\Users\itacag01\AppData\Local\Arduino15\packages\ATTinyCore\hardware\avr\1.5.2\libraries\SoftwareSerial\SoftwareSerial.cpp:121:6: warning: always_inline function might not be inlinable [-Wattributes]
void SoftwareSerial::recv()
^~~~~~~~~~~~~~

when I load the bootloader the warnings are:

***failed;
avrdude: WARNING: invalid value for unused bits in fuse "efuse", should be set to 1 according to datasheet
This behaviour is deprecated and will result in an error in future version
You probably want to use 0x01 instead of 0xff (double check with your datasheet first).

**Reply by SpenceKonde** [reply]
None of those are functional issues, just shitty programming by the people who wrote that, or... really just typical Arduino Maintainer Laziness, where anything that kinda works gets the all clear. I should fix the warnings,, though, That library was pulled in from the official core years ago because people demanded it.  I should fix the warnings though....

I had been expecting the "appears to be a misspelled interrupt handler  one - which generally resuts in the same behavior you were getting through the same mechanism only there the interrupt handler is there, but not linked to the interrupt, whereas in the case of software serial, the ISRs were all  #ifdef'ed out and no ISRs got created at all. Either way, interrupt gets enabled and triggered with no handler, and fails with reset loop.

but yeah, I would think that adding that code would solve your problem until I can get a new version out - right now the other two cores are taking a lot of my time. I've got big things planned and partly done here.... but it's not a small amount of work.

**Comment by DavidAlonsoGalindo** [comment]
Hello again

I made the changes you mentioned and the attiny861 no longer restarts.

Thank you very much for your help, I really appreciate it.

**Reply by DavidAlonsoGalindo** [reply]
Until now I only send and receive data through the serial port, and with the Serial software I only send data, I would need to read data with this same port to verify that there is no problem

**Reply by SpenceKonde** [reply]
Huh, so the builtin Serial implementation worked for receiving for you without anything weird? I think I finally figured out the issue today - #475 (open since november); my understanding of assembly finally got to where I could figure out what appeared to be going wrong and sort-of why the old tricks we knew to fix it works (I it happens when you have no local variables that the compiler can;t optimize away, and this results in the Y pointer being left in as state other than what Serial.available expects and relies on.

Only one port can either send or receive at any time. That is the curse of software serial implementations (the Serial on 861 is a software implementation, - it trades fixed pins for not taking over every single pin interrupt on the entire chip.

**Comment by SpenceKonde** [comment]
Fix is in, located in 2.0.0-dev branch. Simple stuff - warnings should be gone and it now jhas an interrupt vector to ise - you should be able to pull in that file to an older version of ATTinycore and it should work correctly there, too. Please try that and let me know how well it works, I'm not set up to actually test the coder, because I don't want to switch my environment back out of 2.0.0-land, but 2.0.0 doesn't compile anything because ti's boards.txt bears no relationship to the current structure of the core.
